### PR TITLE
Add support for IDP initiated SAML2 SSO when there is a SP qualifier.

### PR DESCRIPTION
**Pull request by dewniMW**
This PR resolves https://github.com/wso2/product-is/issues/6412

### CODE REVIEW

File: components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/processors/IdPInitLogoutRequestProcessor.java [hunk 1]
```diff
@@ -122,10 +124,26 @@ private void init(QueryParamDTO[] queryParamDTOs) {
 
         for (QueryParamDTO queryParamDTO : queryParamDTOs) {
             if (SAMLSSOConstants.QueryParameter.SP_ENTITY_ID.toString().equals(queryParamDTO.getKey())) {
-                this.spEntityID = queryParamDTO.getValue();
+                String issuer = splitAppendedTenantDomain(queryParamDTO.getValue());
+                String issuerQualifier = getSPQualifierExists(queryParamDTOs);
```
Comment (ruwanta) [inline components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/processors/IdPInitLogoutRequestProcessor.java]: Could move 128 to 133 to separate private method "resolveIssuerQualfier", as same logic repeated another place in this PR

Comment (dewniMW) [reply components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/processors/IdPInitLogoutRequestProcessor.java]: Added a new method "resolveIssuerQualfier" to SAMLSSOUtil class.

File: components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/util/SAMLSSOUtil.java [hunk 2]
```diff
@@ -2547,4 +2547,24 @@ public static boolean isSignatureValid(AuthnRequest authnRequest, String querySt
             return false;
         }
     }
+
```
Comment (ruwanta) [inline components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/util/SAMLSSOUtil.java]: We need method level comment, explaining what this method supposed to do, what it expects as parameters, and what it returns.

Comment (dewniMW) [reply components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/util/SAMLSSOUtil.java]: Added method level comment.

File: components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/util/SAMLSSOUtil.java [hunk 3]
```diff
@@ -2547,4 +2547,31 @@ public static boolean isSignatureValid(AuthnRequest authnRequest, String querySt
...
+    /**
+     * Appends service provider qualifier to the issuer if a service provider qualifier is present in the request.
+     *
+     * @param queryParamDTOs query parameters present in the request.
+     * @param issuer service provider entity id present in the request.
+     * @return issuer value with qualifier appended.
+     */
+    public static String resolveIssuerQualifier(QueryParamDTO[] queryParamDTOs, String issuer){
```
Comment (inthirakumaaran) [inline components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/util/SAMLSSOUtil.java]: Formatting issue

Comment (dewniMW) [reply components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/util/SAMLSSOUtil.java]: fixed: https://github.com/wso2-extensions/identity-inbound-auth-saml/pull/280

File: components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/util/SAMLSSOUtil.java [hunk 4]
```diff
@@ -2547,4 +2547,31 @@ public static boolean isSignatureValid(AuthnRequest authnRequest, String querySt
...
+
+        String issuerQualifier = getSPQualifierExists(queryParamDTOs);
+        if (StringUtils.isNotBlank(issuerQualifier)) {
+            return SAMLSSOUtil.getIssuerWithQualifier(issuer, issuerQualifier);
+        } else {
+            return issuer;
+        }
+    }
+
+    private static String getSPQualifierExists(QueryParamDTO[] queryParamDTOs) {
```
[the remark this answers is not in the document]
Comment (dewniMW) [reply components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/util/SAMLSSOUtil.java]: fixed: https://github.com/wso2-extensions/identity-inbound-auth-saml/pull/280

File: components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/validators/IdPInitSSOAuthnRequestValidator.java [hunk 5]
```diff
@@ -25,17 +25,16 @@
 import org.wso2.carbon.identity.sso.saml.dto.QueryParamDTO;
 import org.wso2.carbon.identity.sso.saml.dto.SAMLSSOReqValidationResponseDTO;
 import org.wso2.carbon.identity.sso.saml.util.SAMLSSOUtil;
+import org.wso2.carbon.user.api.UserStoreException;
 import org.wso2.carbon.utils.multitenancy.MultitenantConstants;
 
-
-public class IdPInitSSOAuthnRequestValidator extends SSOAuthnRequestAbstractValidator{
+public class IdPInitSSOAuthnRequestValidator extends SSOAuthnRequestAbstractValidator {
```
Comment (sachiniWettasinghe) [inline components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/validators/IdPInitSSOAuthnRequestValidator.java]: Add class description

Comment (dewniMW) [reply components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/validators/IdPInitSSOAuthnRequestValidator.java]: fixed: https://github.com/wso2-extensions/identity-inbound-auth-saml/pull/292

**Pull request merged**